sympy/matrices/random.py

```diff
@@ -0,0 +1,2343 @@
+    [0, 1, 0],
+    [0, 0, 5]])
+
+    >>> elementary(3, (2,1), 4)  # no more random
+    Matrix([
+    [1, 0, 0],
+    [0, 1, 0],
+    [0, 4, 1]])
+
+    >>> elementary(3, (0,1), None)  # no more random
+    Matrix([
+    [0, 1, 0],
+    [1, 0, 0],
+    [0, 0, 1]])
+
+    Parameters
+    ==========
+    dim : integer
+        dimension of matrix
+    index : tuple(integer, integer) (optional)
+        coordinates ``(i,j)`` of matrix operation
+        if ``i`` equals ``j`` the *elementary matrix*
+        will be of type $\mathbf{M}$,
+        else of type $\mathbf{A}$ or $\mathbf{T}$
+    scalar : symbol (optional)
+        value of elementary entry,
+        defaults to values -1 or 1
+        if **scalar** is **None** the *elementary matrix*
+        will be a transposition $\mathbf{T}$ or identity
+    See Also
+    ========
+    diagonal
+    transposition
+
+    """
+
+    index = index or _sample(range(dim), 2)
+    return super_elementary_matrix(dim, index, scalar)
+
+
+def rotation(dim,
+             index=None,
+             scalar=None):
+    r"""a square matrix n x n of a plane rotation
+
+    Explanation
+    ===========
+    The matrix decribes in n dimensional Euclidian space
+    a rotation of $(i,j)$ plane given by a single rotation square matrix
+    $\mathbf{R}=\mathbf{R(c,s)}$ with entries like the identiy matrix but
+
+    .. math::
+
+        R_{ii} = c, \quad R_{ij} = s, \quad R_{ji} = -s, \quad R_{jj} = c
+
+    with $|c|^2+|s|^2=1$.
+    Usually $s, c$ are $\sin\phi$ and $\cos\phi$ for some angle $\phi$.
+
+    In two dimensions this leads to
+
+    .. math::
+
+        \mathbf{R(c,s)} =
+        \left[\begin{array}{cc} c & s \\ -s & c \end{array}\right].
+
+
+    Examples
+    ========
+    >>> from sympy import sqrt, cos, symbols, eye, I, re, im, expand, simplify
+    >>> from sympy.core.random import seed
+    >>> from sympy.matrices.random import rotation
+    >>> seed(1)
+
+    >>> rotation(3)
+    Matrix([
+    [0, 0, -1],
+    [0, 1,  0],
+    [1, 0,  0]])
+
+    >>> cos_a = sin_a = sqrt(2)/2
+    >>> rotation(3, scalar=cos_a)
+    Matrix([
+    [ sqrt(2)/2, sqrt(2)/2, 0],
+    [-sqrt(2)/2, sqrt(2)/2, 0],
+    [         0,         0, 1]])
+
+    >>> rotation(3, (1, 2), scalar=(cos_a, sin_a))  # no more random
+    Matrix([
+    [1,          0,         0],
+    [0,  sqrt(2)/2, sqrt(2)/2],
+    [0, -sqrt(2)/2, sqrt(2)/2]])
+
+    >>> z = cos_a + sin_a * I
+    >>> rotation(3, (1, 2), scalar=z)  # no more random
+    Matrix([
+    [1,          0,         0],
+    [0,  sqrt(2)/2, sqrt(2)/2],
+    [0, -sqrt(2)/2, sqrt(2)/2]])
+
+    >>> r = rotation(3, scalar=z)
+    >>> r.T * r == eye(3)
+    True
+
+    works with symbols too
+
+    >>> cos_phi = cos(symbols('phi'))
+    >>> r = rotation(3, scalar=cos_phi)
+    >>> r
+    Matrix([
+    [              cos(phi), 0, sqrt(1 - cos(phi)**2)],
+    [                     0, 1,                     0],
+    [-sqrt(1 - cos(phi)**2), 0,              cos(phi)]])
+
```

Comment:
Maybe as a special case it would be good to check if the input expression is a `cos` and then use `sin` in that case.

I'm wondering how this works for all `phi`. Obviously it isn't always true that `sin(phi) == sqrt(1 - cos(phi)**2)`.

Reply:
as long as `phi` is a real number this will hold as a rotation matrix. Even 
```
>>> from sympy.abc import phi
>>> from sympy.matrices.random import *
>>> rotation(3, scalar=phi)
Matrix([
[              phi, 0, sqrt(1 - phi**2)],
[                0, 1,                0],
[-sqrt(1 - phi**2), 0,              phi]])
```
looks fine for me. And for `1 - phi**2<0` this is _complex_ rotation, i.e. here in $SU_3$.

So I am not sure what I missed?

Reply:
I'm just wondering whether it would be better to use `sin(phi)` rather than `sqrt(1 - cos(phi)**2)`. The two are not always equal because sometimes `sin(phi) = -sqrt(1 - cos(phi)**2)`.

Reply:
Both, `sin(phi)` and `-sin(phi)`, define proper rotation matrices `Matrix([[c, s], [-s, c]])` and `Matrix([[c, -s], [s, c]])`. Same for any variable `x` as long as `sqrt(x**2-1)` is real. Here `cos(phi)` is just a example for some `x`. Nothing specific about `cos` or `sin`.

As mentioned, helpful to generate sample rotation matrices with two angels `phi` and `psi` for exercises.
```
>>> simplify(orthogonal(4, spec=(cos(phi), cos(phi))))
Matrix([
[                          cos(phi), -(sqrt(2) + 2)*sqrt(sin(phi)**2)/4,                sqrt(sin(phi)**2)/2, (2 - sqrt(2))*sqrt(sin(phi)**2)/4],
[ (sqrt(2) + 2)*sqrt(sin(phi)**2)/4,                           cos(phi), (-2 + sqrt(2))*sqrt(sin(phi)**2)/4,               sqrt(sin(phi)**2)/2],
[              -sqrt(sin(phi)**2)/2,  (2 - sqrt(2))*sqrt(sin(phi)**2)/4,                           cos(phi), (sqrt(2) + 2)*sqrt(sin(phi)**2)/4],
[(-2 + sqrt(2))*sqrt(sin(phi)**2)/4,               -sqrt(sin(phi)**2)/2, -(sqrt(2) + 2)*sqrt(sin(phi)**2)/4,                          cos(phi)]])

```